router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/H2FailureAccrualFactory.scala

```diff
@@ -0,0 +1,254 @@
+
+    state = Dead
+
+    val timerTask = timer.schedule(duration.fromNow) { startProbing() }
+    reviveTimerTask = Some(timerTask)
+
+    removedForCounter.incr(duration.inMilliseconds.toInt)
+
+    didMarkDead()
+  }
+
+  /**
+   * Called by FailureAccrualFactory after marking an endpoint dead. Override
+   * this method to perform additional actions.
+   */
+  protected def didMarkDead() = {}
+
+  /**
+   * Enter 'Probing' state.
+   * The service must satisfy one request before accepting more.
+   */
+  protected def startProbing() = self.synchronized {
+    state = ProbeOpen
+    cancelReviveTimerTask()
+  }
+
+  /**
+   * Exit 'Probing' state (if necessary)
+   *
+   * The result of the subsequent request will determine whether the factory transitions to
+   * Alive (successful) or Dead (unsuccessful).
+   */
+  private[this] def stopProbing() = self.synchronized {
+    state match {
+      case ProbeOpen =>
+        probesCounter.incr()
+        state = ProbeClosed
+      case _ =>
+    }
+  }
+
+  private[this] def classifyResponse(classifier: H2Classifier, h2ReqRep: H2ReqRep): Try[Response] =
+    classifier.responseClassifier.lift(h2ReqRep) match {
+      case Some(ResponseClass.Successful(_)) =>
+        didSucceed()
+        h2ReqRep.response
+      case Some(ResponseClass.Failed(_)) =>
+        didFail()
+        h2ReqRep.response
+      case None =>
+        h2ReqRep.response match {
+          case Return(rep) =>
+            if (rep.stream.isEmpty) {
+              classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, None))))
+              Return(rep)
+            } else {
+              val stream = rep.stream.onFrame {
+                case Return(f) if f.isEnd =>
+                  classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, Some(Return(f))))))
+                case Throw(e) =>
+                  classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Return((rep, Some(Throw(e))))))
+                case _ => ()
+              }
+              Return(Response(rep.headers, stream))
+            }
+          case Throw(e) =>
+            classifyStream(classifier, H2ReqRepFrame(h2ReqRep.request, Throw(e)))
+            Throw(e)
+        }
+    }
+
+  private[this] def classifyStream(classifier: H2Classifier, h2ReqRepFrame: H2ReqRepFrame): Unit =
+    classifier.streamClassifier(h2ReqRepFrame) match {
+      case ResponseClass.Successful(_) => didSucceed()
+      case ResponseClass.Failed(_) => didFail()
+    }
+
+  def apply(conn: ClientConnection) = {
+    underlying(conn).map { service =>
+      // N.B. the reason we can't simply filter the service factory is so that
+      // we can override the session status to reflect the broader endpoint status.
+      new Service[Request, Response] {
+        def apply(request: Request): Future[Response] = {
+          val param.H2Classifier(classifier) =
+            H2ClassifierCtx.current.getOrElse(param.H2Classifier.param.default)
```

Comment:
ah right.  i'd add a comment there because apparently "Ctx.current" wasn't enough of a clue for me...  just something like "This must be computed on each request to get the the classifier from the request-local context" etc